### PR TITLE
[Form] NumberToLocalizedStringTransformer return int if scale = 0

### LINKED ISSUE #35775: The NumberToLocalizedStringTransformer mix int and float inconsistently.
**Symfony version(s) affected**: 3.4, 4.4, 5.0, ...

**Description**  
When using the NumberType, the reverse transform method does:
```
if (false !== strpos($value, $decSep)) {
    $type = \NumberFormatter::TYPE_DOUBLE;
} else {
    $type = PHP_INT_SIZE === 8 ? \NumberFormatter::TYPE_INT64 : \NumberFormatter::TYPE_INT32;
}
```

Which means, 
- If `1` is submitted, an int should be set.
- If `1.0` is submitted, a float should be set.

But then, a private function round is called and does:
```
private function round($number)
    {
        if (null !== $this->scale && null !== $this->roundingMode) {
            // shift number to maintain the correct scale during rounding
            $roundingCoef = pow(10, $this->scale);
            // string representation to avoid rounding errors, similar to bcmul()
            $number = (string) ($number * $roundingCoef);

            switch ($this->roundingMode) {
                ...
                case self::ROUND_HALF_UP:
                    $number = round($number, 0, PHP_ROUND_HALF_UP);
                    break;
                ...
            }

            $number /= $roundingCoef;
        }

        return $number;
    }
```

Which means, 
- If `1` is submitted with scale `null`, an int will be set.
- If `1` is submitted with scale `0`, a float will be set.
- If `1` is submitted with scale `1`, a float will be set.
- If `1.0` is submitted with scale `null`, a float will be set.
- If `1.0` is submitted with scale `0`, a float will be set.
- If `1.0` is submitted with scale `1`, a float will be set.

It's the same for IntegerToLocalizedStringTransformer since it extends NumberToLocalizedStringTransformer with scale 0.

**Possible Solution**  
Update the private round function to cast the value to int if the scale is `0`.

### PR DESCRIPTION
**Pull request by VincentLanglet**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no 
| Tickets       | Fix #35775
| License       | MIT

### CODE REVIEW

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/NumberToLocalizedStringTransformer.php [hunk 1]
```diff
@@ -282,7 +282,7 @@ private function round($number)
                     break;
             }
 
-            $number /= $roundingCoef;
+            $number = 1 === $roundingCoef ? (int) $number : $number / $roundingCoef;
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/NumberToLocalizedStringTransformer.php]: Can you add some tests please? Do you need help to write them?

Comment (VincentLanglet) [reply src/Symfony/Component/Form/Extension/Core/DataTransformer/NumberToLocalizedStringTransformer.php]: Done, it was easy to add.

$this-> assertSame fail without the following modification. (Failed asserting that -1234.0 is identical to -1234.)
```
$number = 1 === $roundingCoef ? (int) $number : $number / $roundingCoef;
```

Comment (xabbuh) [reply src/Symfony/Component/Form/Extension/Core/DataTransformer/NumberToLocalizedStringTransformer.php]: What about changing the `if` above?

```php
if (null !== $this->scale && 0 !== $scale && null !== $this->roundingMode) {
   // ...
}
```

Comment (VincentLanglet) [reply src/Symfony/Component/Form/Extension/Core/DataTransformer/NumberToLocalizedStringTransformer.php]: @xabbuh  If the user submit the following value `1,23` with the scale `null`, I expect to save the value `1,23`.
If the user submit the following value `1,23` with the scale `0`, I expect to save the rounded value `1`.

If I make your proposed change, i will save the value `1,23` in both cases.

**Review by nicolas-grekas — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @VincentLanglet.

**Pull request merged**